File: Tutorials/C#/Tutorials/Tutorial03.cs
using System;
using System.Windows.Forms;
using TutorialsBase;
using NetOffice;
using Excel = NetOffice.ExcelApi;

namespace TutorialsCS4
{
    public partial class Tutorial03 : UserControl, ITutorial
    {
        #region Fields

        private Excel.Application _application;

        #endregion

        #region Ctor

        public Tutorial03()
        {
            InitializeComponent();
            CreateHandle();
            // update proxy count label
            NetOffice.Core.Default.ProxyCountChanged += delegate (int proxyCount)
            {
                Action<int> update = delegate(int i) { labelProxyCount.Text = i.ToString(); };
                labelProxyCount.Invoke(update, proxyCount);
            };
        }

        #endregion

        #region ITutorial

        public void Run()
        {
            // this tutorial shows you 3 ways in NetOffice to see how many com proxies
            // was currently alive in your application
            //
            // 1.) the property: int NetOffice.Core.ProxyCount
            // 2.) the event: NetOffice.Core.ProxyCountChanged
            // 3.) the events: NetOffice.Core ProxyAdded, ProxyRemoved, ProxyCleared
            //     used from NetOffice.Contribution.Controls.InstanceMonitor

            // Note: Sometimes you may wondering why an instance is disposed.
            // For troubleshooting you can trigger ICOMObject.OnDispose event and see strack trace
        }

        public void Connect(IHost hostApplication)
        {
            HostApplication = hostApplication;
            instanceMonitor1.Factory = NetOffice.Core.Default;
        }

        public void Disconnect()
        {
            if (null != _application)
            {
                _application.Quit();
                _application.Dispose();
                _application = null;
            }
            instanceMonitor1.Factory = null;
        }

        public string Uri
        {
            get { return Program.DocumentationBase + "Tutorial03_EN_CS.html"; }
        }

        public string Caption
        {
            get { return "Tutorial03"; }
        }


        public string Description
        {
            get { return "Observable COM proxies"; }
        }

        public UserControl Panel
        {
            get { return this; }
        }

        #endregion

        #region Properties

        internal IHost HostApplication { get; private set; }

        #endregion

        #region Button Trigger

        private void buttonExcel_Click(object sender, EventArgs e)
        {
            if (null == _application)
            {
                // create application
                _application = new Excel.Application();
                _application.DisplayAlerts = false;
                buttonExcel.Text = "Quit Excel";
                buttonWorkbook.Enabled = true;
                buttonAddins.Enabled = true;
                buttonDisposeChildInstances.Enabled = true;
            }
            else
            {
                // dispose application
                _application.Quit();
                _application.Dispose();
                _application = null;
                buttonExcel.Text = "Start Excel";
                buttonWorkbook.Enabled = false;
                buttonAddins.Enabled = false;
                buttonDisposeChildInstances.Enabled = false;
            }
        }

        private void buttonWorkbook_Click(object sender, EventArgs e)
        {
            // 2 new proxies, the workbooks proxy(implicit) and the new workbook from Add()
            if (null != _application)
                _application.Workbooks.Add();
        }

        private void buttonAddins_Click(object sender, EventArgs e)
        {
            if (null != _application)
            {
                // 1 new enumerator proxy and 1 new proxy for any Addin
                foreach (Excel.AddIn item in _application.AddIns)
                    Console.WriteLine(item.Name);
            }
        }

        private void buttonDisposeChildInstances_Click(object sender, EventArgs e)
        {
            // dispose all child instances from application
            _application.DisposeChildInstances();
        }

        #endregion
    }
}

File: Tutorials/VB/Tutorials/Tutorial03.vb
Imports NetOffice
Imports Excel = NetOffice.ExcelApi
Imports NetOffice.ExcelApi.Enums

Public Class Tutorial03
    Implements ITutorial

    Dim _hostApplication As IHost
    Dim _application As Excel.Application

    Public Sub New()

        InitializeComponent()

        CreateHandle()
        Dim changeHandler As Core.ProxyCountChangedHandler = AddressOf Me.Factory_ProxyCountChanged
        AddHandler Core.Default.ProxyCountChanged, changeHandler

    End Sub

    Public Sub Run() Implements TutorialsBase.ITutorial.Run

        ' this tutorial shows you 3 ways in NetOffice to see how many com proxies
        ' was currently alive in your application
        '
        ' 1.) the property: Int NetOffice.Core.ProxyCount
        ' 2.) the event: NetOffice.Core.ProxyCountChanged()
        ' 3.) the events: NetOffice.Core ProxyAdded, ProxyRemoved, ProxyCleared
        '     used from NetOffice.Contribution.Controls.InstanceMonitor

        ' Note Sometimes you may wondering why an instance Is disposed.
        ' For troubleshooting you can trigger ICOMObject.OnDispose event And see strack trace

    End Sub

    Public Sub Connect(ByVal hostApplication As TutorialsBase.IHost) Implements TutorialsBase.ITutorial.Connect

        _hostApplication = hostApplication
        instanceMonitor1.Factory = NetOffice.Core.Default

    End Sub

    Public Sub Disconnect() Implements TutorialsBase.ITutorial.Disconnect

        If Not IsNothing(_application) Then

            _application.Quit()
            _application.Dispose()
            _application = Nothing

        End If

        instanceMonitor1.Factory = Nothing

    End Sub

    Public ReadOnly Property Caption As String Implements TutorialsBase.ITutorial.Caption
        Get
            Return "Tutorial03"
        End Get
    End Property

    Public ReadOnly Property Description As String Implements TutorialsBase.ITutorial.Description
        Get
            Return "Observable COM proxies"
        End Get
    End Property

    Public ReadOnly Property Panel As System.Windows.Forms.UserControl Implements TutorialsBase.ITutorial.Panel
        Get
            Return Me
        End Get
    End Property


    Public ReadOnly Property Uri As String Implements TutorialsBase.ITutorial.Uri
        Get
            Return FormMain.DocumentationBase & "Tutorial03_EN_VB.html"
        End Get
    End Property


    Private Sub buttonExcel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles buttonExcel.Click

        If (IsNothing(_application)) Then
            ' start application
            _application = New Excel.Application()
            _application.DisplayAlerts = False
            buttonExcel.Text = "Quit Excel"
            buttonWorkbook.Enabled = True
            buttonAddins.Enabled = True
            buttonDisposeChildInstances.Enabled = True
        Else
            ' quit application
            _application.Quit()
            _application.Dispose()
            _application = Nothing
            buttonExcel.Text = "Start Excel"
            buttonWorkbook.Enabled = False
            buttonAddins.Enabled = False
            buttonDisposeChildInstances.Enabled = False
        End If

    End Sub

    Private Sub buttonWorkbook_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles buttonWorkbook.Click

        ' 2 new proxies, the workbooks proxy(implicit) and the new workbook from Add()
        If (Not IsNothing(_application)) Then
            _application.Workbooks.Add()
        End If

    End Sub

    Private Sub buttonAddins_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles buttonAddins.Click

        If (Not IsNothing(_application)) Then
            '1 new enumerator proxy and 1 new proxy for any Addin
            Dim item As Excel.AddIn
            For Each item In _application.AddIns
                Console.WriteLine(item.Name)
            Next item
        End If

    End Sub

    Private Sub buttonDisposeChildInstances_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles buttonDisposeChildInstances.Click

        'dispose all child instances from application
        _application.DisposeChildInstances()

    End Sub

    Private Sub Factory_ProxyCountChanged(ByVal proxyCount As Integer)

        If (labelProxyCount.InvokeRequired) Then
            labelProxyCount.Tag = proxyCount.ToString()
            Dim updateHandler As MethodInvoker = AddressOf Me.UpdateLabel
            labelProxyCount.Invoke(updateHandler)
        Else
            labelProxyCount.Text = proxyCount.ToString()
        End If

    End Sub

    ' its possible the event comes from a different thread, the method is an invoke helper to avoid a CrossThreadException
    Private Sub UpdateLabel()

        labelProxyCount.Text = labelProxyCount.Tag

    End Sub

End Class
